### PR TITLE
Various refactorings to rustc_interface

### PR DESCRIPTION
**Pull request by bjorn3**
This should make it easier to move the driver interface away from queries in the future. Many custom drivers call queries like `queries.global_ctxt()` before they are supposed to be called, breaking some things like certain `--print` and `-Zunpretty` options, `-Zparse-only` and emitting the dep info at the wrong point in time. They are also not actually necessary at all. Passing around the query output manually would avoid recomputation too and would be just as easy. Removing driver queries would also reduce the amount of global mutable state of the compiler. I'm not removing driver queries in this PR to avoid breaking the aforementioned custom drivers.

### CODE REVIEW

File: compiler/rustc_interface/src/passes.rs [hunk 1]
```diff
@@ -640,22 +643,49 @@ pub static DEFAULT_QUERY_PROVIDERS: LazyLock<Providers> = LazyLock::new(|| {
...
+) -> Result<&'tcx GlobalCtxt<'tcx>> {
+    let sess = &compiler.sess;
+
+    rustc_builtin_macros::cmdline_attrs::inject(
+        &mut krate,
+        &sess.psess,
+        &sess.opts.unstable_opts.crate_attr,
+    );
+
+    // FIXME modify krate.attrs in place once #65860 is a hard error.
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/passes.rs:661]: I disagree with this and would like to see everything expanded as a part of the regular expansion phase (that's why I introduced the current setup in https://github.com/rust-lang/rust/pull/108221), that will allow to expand crate-level macro attributes correctly.
If some crate attribute has to be used before that, then it is an exception that should be clearly visible, such attributes need to have some restrictions (e.g. cannot be macro-generated).

Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs]: Reverted

**Comment by bjorn3** [comment]
Seems like `dep_graph.finish_encoding()` needs to be called even if the incr comp session will never be finalized due to a compile error.

**Comment by bjorn3** [comment]
I can't reproduce the bug locally. I've dropped the offending commit for now.

**Comment by michaelwoerister** [comment]
Current state looks good to me. Thanks, @bjorn3!

Let's not put this into a rollup in case it causes unforeseen breakage.
@bors r+ rollup=never

**Pull request merged**